client/src/includes/tabs.js

```diff
@@ -0,0 +1,189 @@
+const DEFAULT_OPTIONS = {
+  tabActiveClass: 'active',
+  paneActiveClass: 'active',
+  tabLinkSelector: '[role="tab"]',
+  initialActiveTab: 0,
+};
+
+class TabInterface {
+  /**
+   * TabInterface expects to be provided an element which is the ancestor of all
+   * tabs the interface includes, most likely an element with role="tablist"
+   *
+   * Tab elements should either be the tab list's immediate children or listed
+   * in the tab list's aria-owns property.
+   *
+   * To specify which tab is active first, ensure that tab has the active class.
+   * Otherwise TabInterface will automatically activate the first tab.
+   *
+   * TabInterface will handle filling in aria-selected, aria-controls,
+   * aria-labelledby, and hidden html attributes. Please ensure that the HTML
+   * contains appropriate role attributes and that each tab has either an href
+   * _or_ an aria-control attribute pointing to the appropriate tab panel.
+   */
+  constructor(el, options) {
+    // If this element has already been instantiated as a tabpanel, bail early
+    if ('wagtailTabInterface' in el) return;
+
+    this.el = el;
+    this.el.wagtailTabInterface = this;
+
+    this.options = Object.assign({}, DEFAULT_OPTIONS, options);
+
+    this.handleKeypress = this.handleKeypress.bind(this);
+    this.setActiveTab = this.setActiveTab.bind(this);
+    this.render = this.render.bind(this);
+
+    this.tabs = [...this.el.querySelectorAll(this.options.tabLinkSelector)];
+
+    this.activeTab = this.tabs[this.options.initialActiveTab];
+
+    this.setUpTabs();
+
+    // If there's a tab that already has the active class, activate it instead
+    const activeClassTab = this.el.querySelector(`${this.options.tabLinkSelector}.${this.options.tabActiveClass}`);
+    if (activeClassTab) {
+      this.activeTab = activeClassTab;
+    }
+
+    // If there's a tab that matches the location hash, activate it instead
+    if (window.location.hash) {
+      const activeHashTab = this.el.querySelector(`[href="${window.location.hash}"]`);
```

Comment:
Note for future reference that the change in https://github.com/wagtail/wagtail/pull/6155 will _not_ need to be implemented here, as this isn’t using jQuery.

But it might still be worth considering "what happens if the location hash is set to something that exists but isn’t a tab" nonetheless.